### PR TITLE
Integrate FIDO2 support

### PR DESCRIPTION
**Pull request by romanstrobl**
This pull request integrates the FIDO2 support in PowerAuth server into the `develop` branch.

### CODE REVIEW

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/converter/serialization/AuthenticatorDataDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.io.Serial;
+import java.nio.ByteBuffer;
+import java.util.Map;
+
+/**
+ * @author Petr Dvorak, petr@wultra.com
```
Comment (banterCZ) [inline powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/converter/serialization/AuthenticatorDataDeserializer.java:43]: dtto

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/entity/AttestationObject.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+package com.wultra.powerauth.fido2.rest.model.entity;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.wultra.powerauth.fido2.rest.model.converter.serialization.AttestationStatementDeserializer;
+import com.wultra.powerauth.fido2.rest.model.converter.serialization.AuthenticatorDataDeserializer;
+import lombok.Data;
+
+/**
+ * @author Petr Dvorak, petr@wultra.com
```
Comment (banterCZ) [inline powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/entity/AttestationObject.java:28]: dtto

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/entity/AuthenticatorData.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+package com.wultra.powerauth.fido2.rest.model.entity;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.PositiveOrZero;
+import lombok.Data;
+
+/**
+ * @author Petr Dvorak, petr@wultra.com
```
Comment (banterCZ) [inline powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/entity/AuthenticatorData.java:27]: dtto

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/entity/CollectedClientData.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.wultra.powerauth.fido2.rest.model.converter.serialization.Base64ToStringDeserializer;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotEmpty;
+import lombok.Data;
+
+/**
+ * @author Petr Dvorak, petr@wultra.com
```
Comment (banterCZ) [inline powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/entity/CollectedClientData.java:29]: dtto

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/request/RegistrationRequest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    @NotBlank
+    private String applicationId;
+    @NotBlank
+    private String activationName;
+    private String expectedChallenge;
+
+    // Authenticator parameters
+    private AuthenticatorParameters authenticatorParameters;
+
+
```
Comment (banterCZ) [inline powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/request/RegistrationRequest.java:41]:
```suggestion

```

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthAuthenticatorProvider.java [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+        authenticatorDetail.setApplicationId(activation.getApplicationId());
+        authenticatorDetail.setUserId(activation.getUserId());
+        authenticatorDetail.setActivationId(activation.getActivationId());
+        authenticatorDetail.setActivationStatus(activation.getActivationStatus());
+        authenticatorDetail.setActivationName(activation.getActivationName());
+        authenticatorDetail.setExternalId(activation.getExternalId());
+        try {
+            authenticatorDetail.setExtras(objectMapper.readValue(activation.getExtras(), new TypeReference<HashMap<String,Object>>() {}));
+        } catch (JsonProcessingException e) {
+            //
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthAuthenticatorProvider.java:268]: ❗ The exception is swallowed.

**Review by banterCZ — APPROVED** [review]
In general, I approve of making development easier by incorporating this feature branch. There is some work to polish the code, but it may be done in consecutive pull requests.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthAuthenticatorProvider.java (hunk 6 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/fido2/PowerAuthAuthenticatorProvider.java]: Already fixed in separate pull request.

File: powerauth-fido2/pom.xml [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+	<dependencies>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-web</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-validation</artifactId>
+			<exclusions>
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-fido2/pom.xml]: Fixed.

File: powerauth-fido2/pom.xml [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+		<dependency>
+			<groupId>org.projectlombok</groupId>
+			<artifactId>lombok</artifactId>
+			<optional>true</optional>
+		</dependency>
+
+		<dependency>
+			<groupId>io.getlime.security</groupId>
+			<artifactId>powerauth-java-crypto</artifactId>
+			<version>1.5.1</version>
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-fido2/pom.xml]: Reflected parent pom version.

File: powerauth-fido2/pom.xml [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+		<dependency>
+			<groupId>org.postgresql</groupId>
+			<artifactId>postgresql</artifactId>
+			<scope>runtime</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.projectlombok</groupId>
+			<artifactId>lombok</artifactId>
+			<optional>true</optional>
+		</dependency>
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-fido2/pom.xml]: Removed.

File: powerauth-fido2/pom.xml [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+			<artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
+			<version>2.1.0</version>
+		</dependency>
+
+		<dependency>
+			<groupId>com.fasterxml.jackson.dataformat</groupId>
+			<artifactId>jackson-dataformat-cbor</artifactId>
+			<version>2.14.2</version>
+		</dependency>
+		<dependency>
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-fido2/pom.xml]: CBOR parsing is not included in PAS.

**Comment by romanstrobl** [comment]
JavaDoc will be added in a separate pull request, there is a lot of it missing right now.

File: powerauth-fido2/pom.xml [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            <exclusions>
+                <exclusion>
+                    <groupId>org.apache.tomcat.embed</groupId>
+                    <artifactId>tomcat-embed-el</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>org.springdoc</groupId>
+            <artifactId>springdoc-openapi-starter-webmvc-ui</artifactId>
```
Comment (banterCZ) [inline powerauth-fido2/pom.xml:50]: `swagger-annotations-jakarta` should be enough.

File: powerauth-fido2/pom.xml [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            <version>2.1.0</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.dataformat</groupId>
+            <artifactId>jackson-dataformat-cbor</artifactId>
+            <version>2.14.2</version>
```
Comment (banterCZ) [inline powerauth-fido2/pom.xml:57]: Version should be taken from the parent.
```suggestion
```

**Review by banterCZ — DISMISSED** [review]
Thanks for improvements of the pom.

File: powerauth-fido2/pom.xml [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-validation</artifactId>
+			<exclusions>
+				<exclusion>
+					<groupId>org.apache.tomcat.embed</groupId>
+					<artifactId>tomcat-embed-el</artifactId>
+				</exclusion>
+			</exclusions>
+		</dependency>
+		<dependency>
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-fido2/pom.xml]: Fixed.

**Review by banterCZ — APPROVED** [review]

**Pull request merged**